src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs

```diff
@@ -210,11 +218,92 @@ private static bool Analyze(
             BoundBlock block,
             DiagnosticBag diagnostics,
             out bool needsImplicitReturn,
-            out ImmutableArray<FieldSymbol> implicitlyInitializedFieldsOpt)
+            out ImmutableArray<FieldSymbol> implicitlyInitializedFieldsOpt,
+            out PooledDictionary<BoundNode, bool> asyncTryFinallyEndsReachable)
         {
-            needsImplicitReturn = ControlFlowPass.Analyze(compilation, method, block, diagnostics);
+            needsImplicitReturn = ControlFlowPass.Analyze(compilation, method, block, diagnostics, out asyncTryFinallyEndsReachable);
             DefiniteAssignmentPass.Analyze(compilation, method, block, diagnostics, out implicitlyInitializedFieldsOpt, requireOutParamsAssigned: true);
             return needsImplicitReturn || !implicitlyInitializedFieldsOpt.IsDefault;
         }
+
+        internal sealed class AsyncTryFinallyEndsReachableRewriter : BoundTreeRewriterWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+        {
+            public static BoundBlock Rewrite(BoundBlock node, PooledDictionary<BoundNode, bool> asyncTryFinallyEndsReachable)
+            {
+                var rewriter = new AsyncTryFinallyEndsReachableRewriter(asyncTryFinallyEndsReachable);
+                Debug.Assert(asyncTryFinallyEndsReachable.All(kvp => kvp.Key is BoundTryStatement
+                    or BoundForEachStatement { AwaitOpt: not null }
+                    or BoundUsingStatement { AwaitOpt: not null }
+                    or BoundUsingLocalDeclarations { AwaitOpt: not null }));
+                return rewriter.Visit(node);
+            }
+
+            private readonly PooledDictionary<BoundNode, bool> _asyncTryFinallyEndsReachable;
+
+            private AsyncTryFinallyEndsReachableRewriter(PooledDictionary<BoundNode, bool> asyncTryFinallyEndsReachable)
+            {
+                _asyncTryFinallyEndsReachable = asyncTryFinallyEndsReachable;
+            }
+
+            private T Visit<T>(T node) where T : BoundNode => (T)base.Visit(node);
+
+            private AsyncTryFinallyEndReachable GetReachability(BoundNode node, AsyncTryFinallyEndReachable originalReachability)
+            {
+                if (_asyncTryFinallyEndsReachable.TryGetValue(node, out var returnValue))
+                {
+                    var reachability = returnValue ? AsyncTryFinallyEndReachable.Reachable : AsyncTryFinallyEndReachable.Unreachable;
+                    Debug.Assert(originalReachability == AsyncTryFinallyEndReachable.Unknown || originalReachability == reachability);
+                    return reachability;
+                }
+                else
+                {
+                    return originalReachability != AsyncTryFinallyEndReachable.Unknown
+                        ? originalReachability
+                        : AsyncTryFinallyEndReachable.Ignored;
```

Comment:
It looks like we are indicating the intent to not specify reachability by not including an entry in the dictionary. I am slightly concerned that this condition is hard to distinguish from not including an entry because of a bug in the visiting logic.